archive/tar.go

```diff
@@ -497,18 +502,32 @@ type ChangeWriter struct {
 	addedDirs map[string]struct{}
 }
 
+// ChangeWriterOpt can be specified in NewChangeWriter.
+type ChangeWriterOpt func(cw *ChangeWriter)
+
+// WithWhiteoutTime sets the whiteout timestamp.
+func WithWhiteoutTime(tm time.Time) ChangeWriterOpt {
+	return func(cw *ChangeWriter) {
+		cw.whiteoutT = tm
+	}
+}
+
 // NewChangeWriter returns ChangeWriter that writes tar stream of the source directory
 // to the privided writer. Change information (add/modify/delete/unmodified) for each
 // file needs to be passed through HandleChange method.
-func NewChangeWriter(w io.Writer, source string) *ChangeWriter {
-	return &ChangeWriter{
+func NewChangeWriter(w io.Writer, source string, opts ...ChangeWriterOpt) *ChangeWriter {
+	cw := &ChangeWriter{
 		tw:        tar.NewWriter(w),
 		source:    source,
-		whiteoutT: time.Now(),
+		whiteoutT: epoch.SourceDateEpochOrNow(),
```

Comment:
I don't think libraries should be reading config from environment variables. The behavior of these functions should be defined by the parameters it takes.

Reply:
I thought reading `$SOURCE_DATE_EPOCH`  by default might be helpful for (non-BuildKit) downstreams, but I can remove it if you prefer

Reply:
It is fine if `ctr` handles it by default, but the env access should be in `cmd` packages in my opinion. I also do not think `containerd` binary should handle it by default. That would mean that for the client using the API (eg. buildkit), the default API behavior is undefined and different for different users(eg. causing invalid cache load).

Reply:
Fixed